tiledb/sm/tile/chunked_buffer.cc

```diff
@@ -0,0 +1,644 @@
+        "Cannot init chunk buffers; Total size must be non-zero."));
+  }
+
+  if (chunk_size == 0) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot init chunk buffers; Chunk size must be non-zero."));
+  }
+
+  buffer_addressing_ = buffer_addressing;
+  chunk_size_ = chunk_size;
+
+  // Calculate the last chunk size.
+  last_chunk_size_ = total_size % chunk_size_;
+  if (last_chunk_size_ == 0) {
+    last_chunk_size_ = chunk_size_;
+  }
+
+  // Calculate the number of chunks required.
+  const size_t nchunks = last_chunk_size_ == chunk_size_ ?
+                             total_size / chunk_size_ :
+                             total_size / chunk_size_ + 1;
+
+  buffers_.resize(nchunks);
+
+  capacity_ = chunk_size_ * (buffers_.size() - 1) + last_chunk_size_;
+
+  return Status::Ok();
+}
+
+Status ChunkedBuffer::init_var_size(
+    const BufferAddressing buffer_addressing,
+    std::vector<uint32_t>&& var_chunk_sizes) {
+  if (!buffers_.empty()) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot init chunk buffers; Chunk buffers non-empty."));
+  }
+
+  if (var_chunk_sizes.empty()) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot init chunk buffers; Var chunk sizes must be non-empty."));
+  }
+
+  buffer_addressing_ = buffer_addressing;
+  var_chunk_sizes_ = std::move(var_chunk_sizes);
+  buffers_.resize(var_chunk_sizes_.size());
+
+  capacity_ = 0;
+  for (const auto& var_chunk_size : var_chunk_sizes_) {
+    if (var_chunk_size == 0) {
+      clear();
+      return LOG_STATUS(Status::ChunkedBufferError(
+          "Cannot init chunk buffers; Var chunk size must be non-empty."));
+    }
+
+    capacity_ += var_chunk_size;
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkedBuffer::alloc_discrete(
+    const size_t chunk_idx, void** const buffer) {
+  if (buffer_addressing_ != BufferAddressing::DISCRETE) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot alloc discrete internal chunk buffer; Chunk buffers are not "
+        "discretely allocated"));
+  }
+
+  if (chunk_idx >= buffers_.size()) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot alloc internal chunk buffer; Chunk index out of bounds"));
+  }
+
+  buffers_[chunk_idx] = std::malloc(get_chunk_capacity(chunk_idx));
+  if (!buffers_[chunk_idx]) {
+    LOG_FATAL("malloc() failed");
+  }
+
+  if (buffer) {
+    *buffer = buffers_[chunk_idx];
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkedBuffer::free_discrete(const size_t chunk_idx) {
+  if (buffer_addressing_ != BufferAddressing::DISCRETE) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot free discrete internal chunk buffer; Chunk buffers are not "
+        "discretely allocated"));
+  }
+
+  if (chunk_idx >= buffers_.size()) {
+    return LOG_STATUS(Status::ChunkedBufferError(
+        "Cannot free internal chunk buffer; Chunk index out of bounds"));
+  }
+
+  ::free(buffers_[chunk_idx]);
```

Comment:
This is actually opposite: I want to invoke the c-api `free()` instead of `this->free()`. I'll just add a comment here for clarity because this syntax is rare.

Reply:
Done